source/server/worker_impl.cc

```diff
@@ -15,16 +13,28 @@ namespace Envoy {
 namespace Server {
 
 WorkerPtr ProdWorkerFactory::createWorker() {
-  return WorkerPtr{new WorkerImpl(tls_, api_.allocateDispatcher())};
+  Event::DispatcherPtr dispatcher(api_.allocateDispatcher());
+  return WorkerPtr{
+      new WorkerImpl(tls_, std::move(dispatcher),
+                     Network::ConnectionHandlerPtr{new ConnectionHandlerImpl(log(), *dispatcher)})};
 }
 
-WorkerImpl::WorkerImpl(ThreadLocal::Instance& tls, Event::DispatcherPtr&& dispatcher)
-    : tls_(tls), dispatcher_(std::move(dispatcher)),
-      handler_(new ConnectionHandlerImpl(log(), *dispatcher_)) {
+WorkerImpl::WorkerImpl(ThreadLocal::Instance& tls, Event::DispatcherPtr&& dispatcher,
+                       Network::ConnectionHandlerPtr handler)
+    : tls_(tls), dispatcher_(std::move(dispatcher)), handler_(std::move(handler)) {
   tls_.registerThread(*dispatcher_, false);
 }
 
 void WorkerImpl::addListener(Listener& listener) {
+  // If the worker thread is already started, post to it. Otherwise do it inline.
+  if (thread_) {
```

Comment:
It has to do with a failure to start listening that can happen in the multi-process case which we want to expose right away. I will add a comment.